trie/snap_diskcache.go

```diff
@@ -0,0 +1,204 @@
+// Copyright 2021 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/VictoriaMetrics/fastcache"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+var (
+	// cacheSizeLimit is the maximum size of the disk cache that aggregates
+	// the writes from above until it's flushed into the disk. Do not
+	// increase the cache size arbitrarily, otherwise the system pause
+	// time will increase when the database writes happen.
+	cacheSizeLimit = uint64(256 * 1024 * 1024)
+)
+
+// diskcache is a collection of dirty trie nodes to aggregate the disk
+// write. It can act as an additional cache to avoid hitting disk too much.
+// diskcache is not thread-safe, callers must manage concurrency issues
+// by themselves.
+type diskcache struct {
+	seq   uint64                 // The number of state transitions contained
+	nodes map[string]*cachedNode // The dirty node set, mapped by storage key
+	size  uint64                 // The approximate size of cached nodes
+}
+
+// newDiskcache initializes the dirty node cache with the given node set.
+func newDiskcache(nodes map[string]*cachedNode, seq uint64) *diskcache {
+	if nodes == nil {
+		nodes = make(map[string]*cachedNode)
+	}
+	var size uint64
+	for key, node := range nodes {
+		size += uint64(node.memorySize(len(key)))
+	}
+	return &diskcache{seq: seq, nodes: nodes, size: size}
+}
+
+// node retrieves the node with given storage key and hash.
+func (cache *diskcache) node(storage []byte, hash common.Hash) (*cachedNode, error) {
+	n, ok := cache.nodes[string(storage)]
+	if ok {
+		if n.hash != hash {
+			owner, path := DecodeStorageKey(storage)
+			return nil, fmt.Errorf("%w %x(%x %v)", errUnexpectedNode, hash, owner, path)
+		}
+		triedbDirtyHitMeter.Mark(1)
+		triedbDirtyNodeHitDepthHist.Update(int64(128))
```

Comment:
Shouldn't this be some `len(DecodeStorageKey(storage)` and not hard coded 128?

Reply:
It's the number of lookups. Since It's already in diskLayer, so just the hard-coded 128.

Reply:
Ah, I thought it's the depth within the trie. (doh, then it would be 32 or 64). Either way, could you please add a small note so we know what this magic number is. Idelaly, would be nice to have this 128 in a constant somewhere :D